File: src/maths/Syracuse.ts
import {MathHelper} from "../helpers/MathHelper";

/**
 * Hypothèse mathématique selon laquelle la suite de Syracuse de n'importe quel entier strictement positif atteint 1.
 */
export class Syracuse {
    private suite: number[] = [];

    constructor(private readonly number: number) {
    }

    private _recursive(n: number): void {
        this.suite.push(n);
        if (n > 1) {
            this._recursive(n % 2 === 0 ? n / 2 : n * 3 + 1);
        }
    }

    generateSuite(): number[] {
        this.suite = [];
        this._recursive(this.number);
        return this.suite;
    }

    get flyTime(): number {
        return this.suite.length - 1;
    }

    get altFlyTime(): number {
        const index = this.suite.findIndex((val) => val < this.number)
        return index !== -1 ? index : this.flyTime;
    }

    get maxAlt(): number {
        return Math.max(...this.suite);
    }
}

export function evaluateSyracuse(n: number) {
    const instances = new Map<number, { flyTime: number; altFlyTime: number; maxAlt: number }>();

    MathHelper.generateRange(n).forEach((num) => {
        const syracuse = new Syracuse(num);
        syracuse.generateSuite();

        instances.set(num, {
            flyTime: syracuse.flyTime,
            altFlyTime: syracuse.altFlyTime,
            maxAlt: syracuse.maxAlt,
        });
    });

    const keys: (keyof { flyTime: number; altFlyTime: number; maxAlt: number })[] = ['flyTime', 'altFlyTime', 'maxAlt'];

    return keys.reduce((acc, key) => {
        const values = Array.from(instances.values())
            .map(item => item[key])
            .filter(value => value !== undefined);

        if (values.length > 0) {
            acc[key] = MathHelper.calculateStats(values);
        }

        return acc;
    }, {} as Record<string, { mean: number; median: number; min: number; max: number }>);
}

File: src/classes/Syracuse.py
from enum import Enum
from typing import List


class Method(Enum):
    NORMAL = "normal"
    COMPRESSED = "compressed"


class SyracuseSuite:
    def __init__(self, suite: List[int]):
        self.suite = suite
        self.initial_value = suite[0]

    @property
    def fly_time(self) -> int:
        return len(self.suite) - 1

    @property
    def alt_fly_time(self) -> int:
        return next((i for i, val in self.suite if val < self.initial_value), self.fly_time)

    @property
    def max_alt(self) -> int:
        return max(self.suite)


class Syracuse:
    @staticmethod
    def _recursive(n: int, method: Method) -> List[int]:
        suite = [n]
        while n > 1:
            if method == Method.NORMAL:
                n = n // 2 if n % 2 == 0 else n * 3 + 1
            elif method == Method.COMPRESSED:
                n = n // 2 if n % 2 == 0 else (n * 3 + 1) // 2
            suite.append(n)
        return suite

    @staticmethod
    def generate_suite(n: int, method: Method = Method.NORMAL, inverse: bool = False) -> SyracuseSuite:
        suite = Syracuse._recursive(n, method)
        return SyracuseSuite(suite)
